A41-xds-rbac.md

```diff
@@ -0,0 +1,230 @@
+server and for which gRPC methods (i.e., whether the client was permitted).
+
+Envoy defines the [RBAC HTTP filter][RBAC filter] that is well-suited for
+limiting which nodes can access which services within a mesh as encouraged by
+the principle of least privilege.
+
+Services that have state managed by their clients tend to need precise
+authorization to that state, commonly driven by per-resource ACLs. RBAC is not
+well suited for this precise authorization, even if the resource name
+is exposed to the engine. Per-resource authorization 1) commonly uses end-user
+identity so requires another filter to perform end-user authentication and
+2) is too great of scale to enumerate all resource's ACLs in a single policy and
+to absorb the rate of change to that policy. Thus RBAC is only intended for
+service-to-service authorization and resource authorization is outside the scope
+of this gRFC.
+
+### Related Proposals:
+
+* [A36: xDS-Enabled Servers][A36]
+* [A29: xDS-Based Security for gRPC Clients and Servers][A29]
+* [A39: xDS HTTP Filter Support][A39]
+
+[A36]: A36-xds-for-servers.md
+[A29]: A29-xds-tls-security.md
+[A39]: A39-xds-http-filters.md
+
+## Proposal
+
+Support the [xDS RBAC HTTP filter][RBAC filter] as a registered filter type on
+xDS-enabled gRPC Servers, and support its `RBACPerRoute` configuration override.
+This does _not_ include [RBAC network filter][] support nor running the filter
+on gRPC clients. xDS v2 support is not necessary.
+
+Supporting the RBAC HTTP filter on server-side leverages [A36: xDS-Enabled
+Servers][A36] and [A39: xDS HTTP Filter Support][A39]. Server-side HTTP filters
+are less common than client-side at this point in time, so this may be the first
+usage of `RouteConfiguration` on server-side and so likely involves adding more
+complete support for A39 by creating and executing filters on server-side and
+processing `VirtualHost`s and `RoutMatch`es to determine which configuration
+should be provided to each filter. A39 should be consulted for the expected
+behavior.
+
+New validation should occur for `HttpConnectionManager` to allow equating
+RBAC's `direct_remote_ip` and `remote_ip`. If the RBAC implementation does not
+distinguish between these fields, then
+`HttpConnectionManager.xff_num_trusted_hops` must be unset or zero and
+`HttpConnectionManager.original_ip_detection_extensions` must be empty. If
+either field has an incorrect value, the Listener must be NACKed. For
+simplicity, this behavior applies independent of the Listener type (both
+client-side and server-side).
+
+The core policy matching logic should be split into an "RBAC engine" to allow
+internal reuse with a non-xDS API. Any non-xDS API will not be RBAC, so this
+reuse is merely an implementation detail. The API is not part of this gRFC.
+There is no requirement that the RBAC engine have a specialized API; it could
+simply be an interceptor.
+
+The [RBAC policy][] has many fields. All current (Envoy-implemented) fields will
```

Comment:
Interesting. We may want a followup change that NACKs if the TypedExtensionConfig is set, simply because having ignored fields in matchers can cause havoc. But I don't think we have much of a need for that feature itself, at least yet.

RBAC here is server-side only, so "upstream IP/port" doesn't make sense. We could have client-side support, but it seems "upstream IP/port" has to be for a more conventional reverse proxy usecase instead of Envoy as a sidecar. I'd have to dig deeper to see if it can work load balancing, but it seems that _specific_ extension is probably not too interesting to us.